dali/tensorflow/dali_dataset.cc

```diff
@@ -0,0 +1,283 @@
+        : DatasetBase(DatasetContext(ctx)),
+          pipe_handles_{pipe_handles},
+          // TODO: Temporary partial shape that will be generalized
+          shapes_{std::vector<PartialTensorShape>(shapes->size(), PartialTensorShape({-1,-1}))},
+          dtypes_{types} {
+    }
+
+    std::unique_ptr<IteratorBase> MakeIteratorInternal(
+        const string& prefix) const override {
+      return std::unique_ptr<IteratorBase>(
+          new Iterator({this, strings::StrCat(prefix, "::DALIDataset")}));
+    }
+
+    const DataTypeVector& output_dtypes() const override {
+      return *dtypes_;
+    }
+
+    const std::vector<PartialTensorShape>& output_shapes() const override {
+      return shapes_;
+    }
+
+    string DebugString() const override { return "DALIDatasetOp::Dataset"; }
+
+   protected:
+    Status AsGraphDefInternal(SerializationContext* ctx,
+                              DatasetGraphDefBuilder* b,
+                              Node** output) const override {
+      return Status::OK();
+    }
+
+  protected:
+    const std::vector<daliPipelineHandle>* pipe_handles_;
+    const std::vector<PartialTensorShape> shapes_;
+    const DataTypeVector* dtypes_;
+
+   private:
+    class Iterator : public DatasetIterator<Dataset> {
+     public:
+      explicit Iterator(const Params& params)
+          : DatasetIterator<Dataset>(params),
+              current_idx_{0},
+              num_pipelines_{dataset()->pipe_handles_->size()} {}
+
+      Status GetNextInternal(IteratorContext* ctx,
+                             std::vector<Tensor>* out_tensors,
+                             bool* end_of_sequence) override {
+          mutex_lock l(mu_);
+
+          if (current_idx_ < num_pipelines_) {
+            auto current_handle = dataset()->pipe_handles_->at(current_idx_);
+
+            daliShareOutput(&current_handle);
+
+            const auto output_count = daliGetNumOutput(&current_handle);
+
+            for (size_t i = 0; i < output_count; i++) {
+              auto shape = DaliToShape(daliShapeAt(&current_handle, i));
+
+              Tensor out(ctx->allocator({}),
+                  dataset()->dtypes_->at(i), shape);
+
+              void *dst = nullptr;
+              switch (dataset()->dtypes_->at(i)) {
+                case DT_FLOAT:
+                      dst = reinterpret_cast<void*>(out.flat<float>().data());
+                  break;
+                case DT_INT32:
+                      dst = reinterpret_cast<void*>(out.flat<int>().data());
+                  break;
+                case DT_INT64:
+                      dst = reinterpret_cast<void*>(out.flat<int64>().data());
+                  break;
+                case DT_HALF:
+                      dst = reinterpret_cast<void*>(out.flat<uint16_t>().data());
+                  break;
+                default:
+                  // todo assert here to catch the failure, because types are not yet converted ?
```

Comment:
We currently support int16 and half as well. Please reuse this from current TF op. I want to have that in sync.

Reply:
Also L242 enforces types. So this `default` should be rather as a sanity check.

Reply:
Was keeping this as a TBD in general, but that might have been confusing/ Now this code has changed, but as you proposed:

1- I added support for new types
2- Added sanity check